src/stdx/unshare.zig

```diff
@@ -0,0 +1,262 @@
+/// subprocesses will have loopback network access only.
+pub fn maybe_unshare_and_relaunch(
+    gpa: std.mem.Allocator,
+    options: struct {
+        pid: bool,
+        network: bool,
+    },
+) !void {
+    assert(builtin.os.tag == .linux);
+
+    const should_unshare_and_fork = blk: {
+        const unshared = std.process.getEnvVarOwned(gpa, "TB_UNSHARED") catch {
+            break :blk true;
+        };
+        gpa.free(unshared);
+        break :blk false;
+    };
+
+    if (should_unshare_and_fork) {
+        try linux_unshare(.{
+            .pid = options.pid,
+            .network = options.network,
+        });
+        if (options.network) {
+            try linux_ip_link_loopback();
+        }
+        try fork_and_exit(gpa);
+    }
+}
+
+/// Implementation of `unshare` somewhat like
+///
+/// ```
+/// unshare --user --net --pid
+/// ```
+///
+/// We're trying to accomplish two main things:
+///
+/// - creating a new pid namespace so that all subprocesses
+///   are automatically terminated when pid 1 (the forked
+///   vortex supervisor) is terminated.
+/// - creating a network sandbox
+///
+/// Note that on recent Ubuntu's this only works if AppArmour
+/// rules have been relaxed:
+///
+/// ```
+/// sudo sysctl -w kernel.apparmor_restrict_unprivileged_unconfined=0
+/// sudo sysctl -w kernel.apparmor_restrict_unprivileged_userns=0
+/// ```
+pub fn linux_unshare(options: struct {
+    pid: bool,
+    network: bool,
+}) !void {
+    assert(builtin.os.tag == .linux);
+
+    // Create user namespace first
+    const unshare_user_result = std.os.linux.unshare(linux.CLONE.NEWUSER);
+    const unshare_user_errno = std.posix.errno(unshare_user_result);
+    if (unshare_user_errno != .SUCCESS) {
+        log.err("Failed to create user namespace: {}", .{unshare_user_errno});
+        return error.UnshareFailure;
+    }
+
+    // Create PID namespace
+    if (options.pid) {
+        const unshare_pid_result = std.os.linux.unshare(linux.CLONE.NEWPID);
+        const unshare_pid_errno = std.posix.errno(unshare_pid_result);
+        if (unshare_pid_errno != .SUCCESS) {
+            log.err("Failed to create pid namespace: {}", .{unshare_pid_errno});
+            return error.UnshareFailure;
+        }
+    }
+
+    // Create network namespace
+    if (options.network) {
+        const unshare_net_result = std.os.linux.unshare(linux.CLONE.NEWNET);
+        const unshare_net_errno = std.posix.errno(unshare_net_result);
+        if (unshare_net_errno != .SUCCESS) {
+            log.err("Failed to create net namespace: {}", .{unshare_net_errno});
+            return error.UnshareFailure;
+        }
+    }
+}
+
+/// Implementation of `ip link` equivalent to
+///
+/// ```
+/// ip link set up dev lo
+/// ```
+///
+/// This brings up the loopback device so that networking
+/// over 127.0.0.1 works.
+pub fn linux_ip_link_loopback() !void {
+    assert(builtin.os.tag == .linux);
```

Comment:
At your discretion can add `comptime assert(builtin.os.tag == .linux);` to underscore that this isn't a runtime check.